core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
             }
+            Expression lowBoundUnwrapped = optionalLowBoundUnwrapped.get();
+            Expression highBoundUnwrapped = optionalHighBoundUnwrapped.get();
+
+            Expression trueIfNotNullExpression = trueIfNotNull(cast.expression());
+            if (trueIfNotNullExpression.equals(lowBoundUnwrapped)) {
+                return highBoundUnwrapped;
+            }
+            if (trueIfNotNullExpression.equals(highBoundUnwrapped)) {
+                return lowBoundUnwrapped;
+            }
+
+            Expression falseIfNotNullExpression = falseIfNotNull(cast.expression());
+            if (falseIfNotNullExpression.equals(lowBoundUnwrapped) || falseIfNotNullExpression.equals(highBoundUnwrapped)) {
+                if (falseIfNotNullExpression.equals(lowBoundUnwrapped) && falseIfNotNullExpression.equals(highBoundUnwrapped)) {
+                    return falseIfNotNullExpression;
+                }
+                return and(lowBoundUnwrapped, highBoundUnwrapped);
+            }
+            Expression castNullToBoolean = new Cast(new Constant(UNKNOWN, null), BOOLEAN);
+            if (castNullToBoolean.equals(lowBoundUnwrapped) || castNullToBoolean.equals(highBoundUnwrapped)) {
+                if (castNullToBoolean.equals(lowBoundUnwrapped) && castNullToBoolean.equals(highBoundUnwrapped)) {
+                    return castNullToBoolean;
+                }
+                return and(lowBoundUnwrapped, highBoundUnwrapped);
+            }
+
+            if (lowBoundUnwrapped instanceof Comparison lowBoundUnwrappedComparison &&
+                    highBoundUnwrapped instanceof Comparison highBoundUnwrappedComparison) {
+                Type sourceType = cast.expression().type();
+                Type lowBoundType = lowBoundUnwrappedComparison.right().type();
+                Type highBoundType = highBoundUnwrappedComparison.right().type();
+
+                if (cast.expression().equals(lowBoundUnwrappedComparison.left()) &&
+                        Objects.equals(sourceType, lowBoundType) &&
+                        cast.expression().equals(highBoundUnwrappedComparison.left()) &&
+                        Objects.equals(sourceType, highBoundType)) {
+                    if (Set.of(GREATER_THAN, GREATER_THAN_OR_EQUAL).contains(lowBoundUnwrappedComparison.operator()) &&
+                            Set.of(LESS_THAN, LESS_THAN_OR_EQUAL).contains(highBoundUnwrappedComparison.operator())) {
+                        // Try to reconstruct the BETWEEN predicate with the cast unwrapped
+                        if (!(lowBoundUnwrappedComparison.right() instanceof Constant(Type _, Object lowBoundValue))
+                                || !(highBoundUnwrappedComparison.right() instanceof Constant(Type _, Object highBoundValue))) {
+                            return expression;
+                        }
+
+                        int compareLowBoundValueAndHighBoundValue = compare(sourceType, lowBoundValue, highBoundValue);
```

Comment:
The `compare` method would throw exception if `lowBoundValue` or `highBoundValue` was null. 

Again, we are making implicit assumptions about what the `unwrapCast` method returns.